File: java/0121. Best Time to Buy and Sell Stock.java
package leetcode.java;

/**
 * Obvious implementation
 * Time complexity: O(N^2)
 * Space complexity: O(1)   
 */
class Solution {
    public int maxProfit(int[] prices) {
        int maxProfit = 0;

        for (int buyIdx=0; buyIdx<prices.length; buyIdx++) {
            for (int sellIdx=buyIdx+1; sellIdx<prices.length; sellIdx++) {
                maxProfit = Math.max(maxProfit, prices[sellIdx] - prices[buyIdx]);
            }
        }
        return maxProfit;
    }
}


/**
 * Efficient implementation
 * Time complexity: O(N)
 * Space complexity: O(1)   
 */
class Solution {
    public int maxProfit(int[] prices) {
        
        int lowest = prices[0];
        int maxProfit = 0;
        for (int price : prices) {
            if (price > lowest) {
                maxProfit = maxProfit > price - lowest ? maxProfit : price - lowest;
            } else {
                lowest = price;
            }
        }
        return maxProfit;
    }
}

File: python/0121. Best Time to Buy and Sell Stock.py
'''
Obvious implementation
Time Complexity: O(N^2)
Space complexity: O(1)
'''
class Solution(object):
    def maxProfit(self, prices):
        maxProfit = 0
        for buyIdx in range(len(prices)):
            for sellIdx in range(buyIdx+1, len(prices)):
                maxProfit = max(maxProfit, prices[sellIdx] - prices[buyIdx])
        return maxProfit


'''
Efficient implementation
Time Complexity: O(N)
Space complexity: O(1)
'''
class Solution(object):
    def maxProfit(self, prices):
        currLow = prices[0]
        maxProfit = 0
        for p in prices[1:]:
            if p < currLow:
                currLow = p
            else:
                maxProfit = max(maxProfit, p - currLow)
        return maxProfit